tests/test_xml_parser_main_metadata_together.c

```diff
@@ -262,37 +263,232 @@ test_cr_xml_parse_main_metadata_together_08_long_primary(void)
     g_assert_cmpint(parsed, ==, 2);
 }
 
+static void
+test_cr_xml_package_iterator_00(void)
+{
+    int parsed = 0;
+    GError *tmp_err = NULL;
+    cr_Package *package = NULL;
+
+    cr_PkgIterator *pkg_iterator = cr_PkgIterator_new(
+        TEST_REPO_02_PRIMARY, TEST_REPO_02_FILELISTS, TEST_REPO_02_OTHER, NULL, NULL, NULL, NULL, TRUE, &tmp_err);
+
+    while ((package = cr_PkgIterator_parse_next(pkg_iterator, &tmp_err))) {
+        parsed++;
+        cr_package_free(package);
+    }
+
+    g_assert(cr_PkgIterator_is_finished(pkg_iterator));
+    cr_PkgIterator_free(pkg_iterator, &tmp_err);
+
+    g_assert(tmp_err == NULL);
+    g_assert_cmpint(parsed, ==, 2);
+}
+
+static void
+test_cr_xml_package_iterator_01_out_of_order_pkgs(void)
+{
+    int parsed = 0;
+    GError *tmp_err = NULL;
+    cr_Package *package = NULL;
+
+    cr_PkgIterator *pkg_iterator = cr_PkgIterator_new(
+        TEST_REPO_02_PRIMARY, TEST_DIFF_ORDER_FILELISTS, TEST_REPO_02_OTHER, NULL, NULL, NULL, NULL, FALSE, &tmp_err);
+
+    while ((package = cr_PkgIterator_parse_next(pkg_iterator, &tmp_err))) {
+        parsed++;
+        cr_package_free(package);
+    }
+    cr_PkgIterator_free(pkg_iterator, &tmp_err);
+
+    g_assert(tmp_err != NULL);
+    g_assert_cmpint(tmp_err->code, ==, CRE_XMLPARSER);
+    g_clear_error(&tmp_err);
+
+    /////////////////
+
+    pkg_iterator = NULL;
+    parsed = 0;
+
+    pkg_iterator = cr_PkgIterator_new(
+        TEST_REPO_02_PRIMARY, TEST_DIFF_ORDER_FILELISTS, TEST_REPO_02_OTHER, NULL, NULL, NULL, NULL, TRUE, &tmp_err);
+
+    while ((package = cr_PkgIterator_parse_next(pkg_iterator, &tmp_err))) {
+        parsed++;
+        cr_package_free(package);
+    }
+    cr_PkgIterator_free(pkg_iterator, &tmp_err);
+
+    g_assert(tmp_err == NULL);
+    g_assert_cmpint(parsed, ==, 2);
+}
+
+// static void
+// test_cr_xml_parse_main_metadata_together_03_newpkgcb_returns_null(void)
```

Comment:
Which of these tests do you think needs to be duplicated for package iterator, since the `parse_xml_together` tests give us some coverage of it already?  And any others else that might be missing?

Reply:
Right I think we don't need to duplicate all the tests, it looks like you already have the most important ones.

I would maybe add a couple of test for invalid input, such as calling `cr_PkgIterator_prase_next` and `cr_PkgIterator_is_finished` with `NULL` and perhaps `cr_PkgIterator_new` with one of the filepaths set to `NULL`?